### ISSUE TITLE
Add a customizable controller button to exit Ryujinx Ingame

### ISSUE THREAD

**Issue by jojodunet** [Ryujinx/Ryujinx#3725]
## Feature Request
- Add a customizable controller button to exit Ryujinx Ingame

#### Overview:
- Add a customizable controller button to exit Ryujinx Ingame (one selectable button to emulate the Alt+F4 keyboard).

#### Smaller Details: 
- Adding in the controller menu the possibility to set a button to exit Ryujinx Ingame with the controller. For an example, for an XBOX controller, the button 7 is free (see picture) when Ryujinx is running and I would like to set the Alt+F4 function to it, to exit Ryujinx ingame (to exit the game and Ryujinx at the same time).
![images](https://user-images.githubusercontent.com/65983141/192304494-70ba67a5-023f-4861-8900-ae44db15761c.png)

#### Nature of Request:  
- Addition in the controller menu/configuration.

### Why would this feature be useful?
- For example, with this addition we can close Ryujinx ingame without the keyboard (it will be useful for an EmulationStation build without keyboard, for example, and for the others frontend applications).

**Comment by Squall-Leonhart** [comment]
Button 7 cannot be assumed to be free.

**Comment by jojodunet** [comment]
The button 7 for XBOX controller it is just an example, It could be another button or buttons combination (Start+Select, for example).
Lot of controllers, other than XBOX controller, has addressable buttons.

**Comment by e2zippo** [comment]
This is probably better of being left to the frontend side, otherwise it would have to be implemented in every emulator you're using. It's already possible to to exactly this in Launchbox. For example. I've mapped R1 + Select on my DualSense to "Alt+F4. If you're using EmulationStation perhaps it's better to ask them to do a similar implementation like Launchbox has?

**Comment by jojodunet** [comment]
I thank you for your interest.

I have already asked at the ES-DE devs team and I got the below reply.
I have also thinking using Steam Overlay to kill the Ryujinx process but it’s no way either.

Here the ES-DE devs team explanation :

« When you launch an emulator or game ES-DE turns over the control to that process, what happens next is either up to that application or the operating system.
Under normal circumstances ES-DE is even completely suspended (unless you run the emulator while keeping ES-DE running in the background which is required by some emulators or games)
If running ES-DE via Steam on the Steam Deck then emulators can be closed using the Steam overlay which will return focus to ES-DE, but I suspect that requires OS-level integration to work as this is a custom operating system and it's not done by ES-DE anyway. »

**Comment by Shituation** [comment]
https://github.com/Ryujinx/Ryujinx/issues/3848

**Comment by jlar0che** [comment]
This feature is vital. I'm on Steam Deck using Ryujinx under EmulationStation-DE. Can't exit out of a game without pressing the Steam button and then exiting out of the entire EmulationStation along with Ryujinx. Then I have to fire up EmulationStation again and navigate back to my Switch library to continue playing Switch games with Ryujinx.

In many ways I find Yuzu to be inferior, but at least I can exit out of it using the "Start" + "Select" key combo.

**Comment by Shituation** [comment]
>
> 
> In many ways I find Yuzu to be inferior, but at least I can exit out of it using the "Start" + "Select" key combo.

Just being genuinely curious, why do you find yuzu inferior? I'm using it after realizing ryujinx doesn't have basic gamepad navigation and found it to be much more mature than the last time I tried it, it has vulcan and gamepad shortcuts... even the guide Xbox button can be used (if you use it on pc) which most (if not all) software besides steam and MS ignore.

**Comment by TomTurbine** [comment]
Can support this.

I have a PS5 controller and would love to be able to use the Touchpad, Mic, and PS Logo buttons.

Want to have the logo bound to exit the emulator. While the Touchpad is  potentially 2 other buttons while the Mic Mute is another potential hot key.

**Comment by Shituation** [comment]
> Can support this.
> 
> I have a PS5 controller and would love to be able to use the Touchpad, Mic, and PS Logo buttons.
> 
> Want to have the logo bound to exit the emulator. While the Touchpad is potentially 2 other buttons while the Mic Mute is another potential hot key.

You already can with Yuzu. I don't understand why people are favoring ryujinx over yuzu.
I came here the first time because I didn't know yuzu supported all the basic features that ryujinx is missing.

**Comment by edisionnano** [comment]
> > Can support this.
> > I have a PS5 controller and would love to be able to use the Touchpad, Mic, and PS Logo buttons.
> > Want to have the logo bound to exit the emulator. While the Touchpad is potentially 2 other buttons while the Mic Mute is another potential hot key.
> 
> You already can with Yuzu. I don't understand why people are favoring ryujinx over yuzu. I came here the first time because I didn't know yuzu supported all the basic features that ryujinx is missing.

Thanks for your totally useful comment

**Comment by Shituation** [comment]
¯\_(ツ)_/¯

keep begging then. 

Salut!

**Comment by jhildenbiddle** [comment]
> This is probably better of being left to the frontend side, otherwise it would have to be implemented in every emulator you're using.

This functionality is available in every other emulator I'm using (Retroarch and RPCS3) except for Ryujinx. RetroArch allows exiting directly from the controller or via an overlay menu. RPCS3 does not (as far as I know) allow exiting directly from the controller, but it [recently added](https://github.com/RPCS3/rpcs3/pull/13226) an overlay menu with the ability to exit games.

![CleanShot 2023-02-03 at 11 55 54@2x](https://user-images.githubusercontent.com/442527/216699240-9100cb24-61f8-4209-9191-0b0bd7222041.png)

The ability to exit via the controller is how I am able to use RetroArch and RCPS3 with [ES-DE](https://es-de.org) running as a set-top box. It's great. Since Ryujinx does not offer a similar "exit via controller" feature, it is the one emulator that requires 1) the use of a keyboard or 2) additional software to map keyboard commands to controller buttons.

My hope is--as @jojodunet recommended above--that the button most typically used to display a console's overlay menu (e.g. Xbox = the center Xbox button on Xbox controllers) would be used by default to pull up a Ryujinx overlay menu similar to what RPCS3 is doing. This is the last piece of the puzzle I need to run a macOS-based set-top emulation box running Ryujinx.

Thanks for considering the feature, and thanks for all of the truly amazing work so far. Ryujinx is mind-blowing and I love it.

**Comment by MissileMax** [comment]
Please, somebody,  add this feature! I use Playnite, and it is a pain to find a keyboard just to exit the game.

**Comment by Kupo91** [comment]
I'm using Pegasus Frontend and wish to have this feature, too. As you see the request is not for a specific frontend, it's more for couch gaming in general.

Duckstation and pcsx2 offer a big picture UI, which can be opened via custom buttons and which gives you access to several options, including exiting the emulator. Dolphin and yuzu have no big picture UI, but atleast let you map a button to exit the emulator. It's currently the main reason why I use yuzu over ryujinx.

**Comment by TomTurbine** [comment]
And as an extension of the post above this.

Yuzu also has the functionality using the Home button and the Minus button combination to exist.

**Comment by BENETNATH** [comment]
would be really useful !

**Comment by lolekuk** [comment]
Yes, please. So for example, when booting emulator from retropie/emulationstation we can exit games back to retropie/emulationstation games dashboard.

**Comment by m33ts4k0z** [comment]
Please add this feature. Its a very important feature that every emulator has. Yuzu had it too.

**Comment by zhen-huan-hu** [comment]
Controller combo to exit would be very nice for Steam Deck.

**Comment by lolekuk** [comment]
Please :)

I had emulation station configured to start games from emulation station. Could exit yuzu and go back to emulation station with pad shortcut. Can't do this with ryujinx :(

**Comment by Tsury** [comment]
This is the only thing preventing me from being able to do everything from the controller, please add this.

**Comment by Kupo91** [comment]
PR that would close this issue: https://github.com/Ryujinx/Ryujinx/pull/4503

**Comment by knoopx** [comment]
Here's a dirty patch that maps `Guide` button to exit the emulator for those who launch ryujinx through frontends: https://github.com/knoopx/Ryujinx/commit/2614ca89786d0f6853fe59773b80aa0f1f03e2d4

**Comment by octypus** [comment]
@knoopx How do I apply this?

**Comment by knoopx** [comment]
> @knoopx How do I apply this?

Either clone upstream and apply the patch via git or clone my fork (already outdated) and build following official instructions. I build ryujinx trough nix, ping me if you want me to share the nix package.

**Comment by zhen-huan-hu** [comment]
> Here's a dirty patch that maps `Guide` button to exit the emulator for those who launch ryujinx through frontends: [knoopx@2816b4c](https://github.com/knoopx/Ryujinx/commit/2816b4ce34da99050ef7cce5e710b5005b0dffb1)

Guide bottom does not exist on SteamDeck unfortunately.

**Comment by zhen-huan-hu** [comment]
> PR that would close this issue: #4503

This is the way.
